EVPN Multihoming part-2 - NHG ZAPI Infrastructure and Sharpd Implementation

sworleys opened the pull request:
These patches add new ZAPI definitions for proto-owned NHGs that are sent to zebra via and ID and `struct nexthop_group`.
We have divided the ID space among route_types while maintaining and ID space for zebra-created NHGs and L2-NHGs as well. These patches allow us to make use of the Replace API in the linux kernel for NHGs and are a starting point for route scalability. EVPN-MH will make use of this new API in a later patch set.

Currently this only supports upperlevel protos sending zebra fully resolved `connected` or `onlink` nexthops in the nexthop group. Nexthop resolution is a much harder headache to solve later on given the complexity of route-maps and will have to come later.


To test this functionality you can do this:


```
alfred# c                                                                                                              
alfred(config)# nexthop-group red                                                                                      
alfred(config-nh-group)# nexthop 1.1.1.1 dummy1                                                                        
alfred(config-nh-group)# nexthop 1.1.1.2 dummy2                                                                        
alfred(config-nh-group)# end                                                                                           
alfred# sharp install routes 3.3.3.1 nexthop-group red 1                                                               
alfred# c                                                                                                              
alfred(config)# nexthop-group red                                                                                      
alfred(config-nh-group)# nexthop 1.1.1.3 dummy3                                                                        
alfred(config-nh-group)# no nexthop 1.1.1.1 dummy1                                                                     
alfred(config-nh-group)# end                                                                                           
alfred# 
```

In the above we are directly modifying the nexthop group by adding/removing a nexthop. You can do this with N routes and it will only need to send a single REPLACE operation per change in the linux kernel rather than a route removal and add of the N routes:

```
sworley@alfred ~/D/c/e/frr-2> ip ro show 3.3.3.1
3.3.3.1 nhid 191666659 proto 194 metric 20 
        nexthop via 1.1.1.2 dev dummy2 weight 1 
        nexthop via 1.1.1.3 dev dummy3 weight 1 
```


A hidden command was also added: `zebra nexthop proto only` that will allow users to ONLY create these types of nexthops in the dataplane and none of the ones implicitly created by zebra. This will allow people with various dataplanes to still test NHGs without committing to them fully just yet.

Since these are coming in a separate PR from the EVPN-MH tests, I will add some topotests with sharpd as well before its merged.

TODO:
- [x] Add topotests!

sworleys commented:
Added a topotest

sworleys commented:
poke @mjstapp any comments here?

mjstapp commented on lib/log.c:
```diff
@@ -444,7 +444,9 @@ static const struct zebra_desc_table command_types[] = {
-	DESC_ENTRY(ZEBRA_OPAQUE_UNREGISTER)};
+	DESC_ENTRY(ZEBRA_OPAQUE_UNREGISTER),
+	DESC_ENTRY(ZEBRA_NHG_ADD),
+	DESC_ENTRY(ZEBRA_NHG_DEL)};
```
doesn't the NOTIFY_OWNER message want to be in here too?

sworleys replied:
yep adding

sworleys replied:
added it

mjstapp commented on lib/zclient.h:
```diff
@@ -356,6 +360,7 @@ struct zclient {
 #define ZAPI_MESSAGE_SRCPFX   0x20
 /* Backup nexthops are present */
 #define ZAPI_MESSAGE_BACKUP_NEXTHOPS 0x40
+#define ZAPI_MESSAGE_NHG 0x80
```
so this is using up the last bit in the 'message' flags octet - should at least be worth a comment?

sworleys replied:
What should I comment exactly? api.message is a uint32_t so it has some room left still for more types.

mjstapp commented on lib/zclient.c:
```diff
@@ -967,6 +967,51 @@ int zapi_nexthop_encode(struct stream *s, const struct zapi_nexthop *api_nh,
+
+static void zclient_nhg_writer(struct stream *s, uint16_t proto, int cmd,
+			       uint32_t id, size_t nhops,
+			       struct zapi_nexthop *znh)
```
Hmm - can you say something about why this doesn't follow the common pattern of using a public zapi struct that's used for marshalling and un-ing? There are some benefits to that - a single place to put key validation logic, like "nhops must be < MULTIPATH". without that, it seems like every daemon who wants to work with nhgs will have to have its own copy of validation and error-handling logic. Those structs are also sort of self-documenting - if we need to add something to the zapi nhg, there'd be one place that defined that - for example, when backup nexthops need to be added.

sworleys replied:
So im not sure I follow this comment? I think we are doing it pretty similarly to how its done for routes? Also, its worth noting that each daemons notion of a `nexthop/nexthop group` is very different between protos. For example, sharpd and bgp (the only users of this API for right now) dont share common structs that could call into this code. The way sharpd turns something into a nexthop group is different from how bgp does it.

mjstapp commented on lib/zclient.h:
```diff
@@ -611,6 +625,22 @@ struct zclient_options {
+#define ZEBRA_NHG_PROTO_LOWER                                                  \
+	(ZEBRA_NHG_PROTO_SPACING * (ZEBRA_ROUTE_CONNECT + 1))
+
+extern uint32_t zclient_get_nhg_start(uint32_t proto);
```
So the idea is that each daemon gets a chunk of number-space to use - that's so that zebra can just use the daemons' numbers directly, I guess? But the api here never tells the daemon the actual range that's available to it? I wonder - why not just have zebra manage the 'shared' number space - the daemon could send whatever number it wanted, and zebra could manage the mapping between the 'owning' daemon's number and the shared/kernel/fib number?

sworleys replied:
I dont think I agree with the conversion. There is around 8million nexthops allocated per protocol which is plenty.

I'd rather have the ID bgp thinks its using the be same one being used by zebra. And its up to bgp to manage how it uses that ID space.

> But the api here never tells the daemon the actual range that's available to it
Zebra tells daemons their starting ID.

Given the daemons their own ID space significantly reduces complexity in zebra. And we dont have to worry about negotiating an ID or races on allocating one between daemons.

With these kind of nexthops, zebra really just makes sure they remain in the kernel and its database and does any replace operations. Zebra doesn't do much beyond that. Other daemons cant hash to them and use them so no issues with a proto wanting to remove/change one another is using.

mjstapp commented on sharpd/sharp_zebra.c:
```diff
@@ -351,8 +351,33 @@ void vrf_label_add(vrf_id_t vrf_id, afi_t afi, mpls_label_t label)
+	struct nexthop *nh;
+
+	for (ALL_NEXTHOPS_PTR(nhg, nh)) {
+		api_nh = &nh_array[nexthop_num];
```
So this is the sort of thing I was concerned about in the comment about lib/zclient: limits check for 'nh_array', something that would have to be maintained in every daemon who wants to use the nhg zapi.

sworleys replied:
Addressed in previous Q. I think its pretty similar to how we are sending routes in sharpd unless I misunderstood your statement.

mjstapp replied:
the issue is that this will overrun nh_array if it's not limit-checked. if we're doing something like that in zapi route handling somewhere, it should be fixed too, but it's hard to see adding a new instance of it?

sworleys replied:
ah i see. yes this needs a check.

But I still stand by my other statement related to this that you cant really have a common entry for the converting in to zapi encodings since every daemons idea of a nexthop is different. They are all going to have to encode to zapi_nexthop from their own structs they are using to infer nexthops. I'm pretty sure sharpd is the only one actually using `lib/nexthop_goup`

mjstapp replied:
> you cant really have a common entry for the converting in to zapi encodings

I just don't see that - we do it for routes and lsps, for example, among many other objects, and both of those include nexthops? daemons are, as you say, responsible for converting from their own internal datastructs to the neutral zapi objects. but there are a couple of key benefits: there is a single copy of the marshall/unmarshall code, and there is some reasonable way to estimate message buffer/stream size. another benefit is that the zclient api uses the neutral zapi struct - it doesn't have to be updated every time any new field is added to the struct.

sworleys replied:
If we did that we could have a common function for `nexthop_group ->zapi_nexthop`  but every protocol would still need a conversion from `proto independent stuff -> nexthop_group` right? You think that's worth it?

mjstapp replied:
I ... guess I think the pattern in use for routes and lsps ought to be the default pattern, unless there's some compelling reason not to follow it. I wouldn't personally suggest trying to use `struct nexthop_group` - that's got a specific meaning for configuration. like the route and lsp zapis, I'm suggesting defining a `zapi_nhg` (or something) struct, and using it in the zclient marshall/unmarshall apis. as you say, `zapi_nhg` should contain some `zapi_nexthop` objects, and we already have code that zapis them, so that's reusable.

mjstapp commented on sharpd/sharp_nht.c:
```diff
@@ -65,3 +69,149 @@ void sharp_nh_tracker_dump(struct vty *vty)
+	return NULL;
+}
+
+static void sharp_nhgroup_add_cb(const char *name)
```
so ... this looks like every nhg will get an id assigned to it - is that right? If so, it looks as if that will force all sharp use of nhgs to use the ids only, and that will break things that ... other folks use (like backups?)

sworleys replied:
When you create the nexthop-group, it will send it to zebra. If it fails to install it because its not a type we support or for some other reason, sharpd gets told and reverts to just using the traditional way of sending all nexthops down with the route, rather than ID.

sworleys commented:
Added requested log command type and fixed ALOT of conflicts.

sworleys commented:
ci:rerun

sworleys commented:
ci:rerun

mjstapp commented on lib/nexthop_group.c:
```diff
@@ -820,15 +831,15 @@ static bool nexthop_group_parse_nexthop(struct nexthop *nhop,
-					    nhh->nhvrf_name, nhh->labels, NULL,
-					    nhh->weight, nhh->backup_str));
+	return (nexthop_group_parse_nexthop(
+		nhop, nhh->addr, nhh->intf, nhh->onlink, nhh->nhvrf_name,
```
style: at least here in the webui, it looks like this re-indenting has moved these args more than one indent-level to the left. my editor, at least, won't maintain that: isn't our style that the continuation is just one level "left" after an empty open-paren?

sworleys replied:
just double checked, this seems to be how clang-format likes it which is our style God.

mjstapp commented on lib/route_types.txt:
```diff
@@ -84,7 +84,7 @@ ZEBRA_ROUTE_PBR,        pbr,       pbrd,   'F', 1, 1, 0,     "PBR"
 ZEBRA_ROUTE_OPENFABRIC, openfabric, fabricd,  'f', 1, 1, 1, "OpenFabric"
 ZEBRA_ROUTE_VRRP,       vrrp,      vrrpd,  '-', 0, 0, 0,     "VRRP"
-ZEBRA_ROUTE_NHG,        nhg,       none,  '-', 0, 0, 0,     "Nexthop Group"
+ZEBRA_ROUTE_NHG,        zebra,     none,  '-', 0, 0, 0,      "Nexthop Group"
```
why change this - is "nhg" ... unclear, or ambiguous?

sworleys replied:
to make it clear these are no proto-owned NHGs. These are ones created by zebra from normal routes.

I could make it zebra-nhg but that seemed a little weird.

mjstapp commented on zebra/zapi_msg.c:
```diff
@@ -1689,100 +1690,215 @@ static void zread_route_add(ZAPI_HANDLER_ARGS)
+	 * Resolution is going to need some more work.
+	 */
+	if (nhe)
+		nhg_notify(proto, client->instance, id, ZAPI_NHG_INSTALLED);
```
mmm - we shouldn't tell the owner/daemon the result until we have the result. we have a results callback that is being used - can the nhg results be moved there, to align with what we do for route redist and nht?

sworleys replied:
sure that makes sense. This was just temp for now since we are only doing onlink and they are expected to always be success.

We will still may need to notify fail here since it can early fail in the add function if it doesnt meet a few params.

sworleys replied:
Hmm.. so this is a bit more complex than I had thought originally. We have to do instance tracking since there is the possibility of having multiple instance of the protos. It will require a lot of code changes in the NHE and dataplane code to address this from the looks of it to get it closer to the way routes track it.

Since this is a fairly basic use case for now. We arent doing resolution or much with these proto nexthops yet, zebra really just sends them off to the dataplane. I'd like to postpone adding that code until next iteration so we can get this patch in. Though I definitely agree it should be moved to consolidate into that path since the way it is right now was just to simplify it.

mjstapp replied:
but ... doesn't this contradict what you wrote when I asked about the sharpd behavior above? `zebra_nhg_proto_add()` is going to be able to allocate an nhg/nhe - and this test will cause zebra to always send sharpd an INSTALLED notification. that will cause sharpd to only use the "installed" nhgs - but though the way you've defined the zapi messaging and the apis doesn't include backup nexthops, so sharpd will no longer be able to test backups.

sworleys replied:
it would only mark success if it met the condition that its onlink which should just install success anyway.

Though I do see your concern with backup nexthop objects. I'm not really willing to support those just yet until i figure out how to handle resolution. I think the support will come naturally through that.

Would you be okay with me failing if a backup nexthop has been given for now? So it will always revert to the old way with those then?

mjstapp commented on zebra/zebra_nhg.h:
```diff
@@ -249,13 +260,50 @@ extern int zebra_nhg_kernel_find(uint32_t id, struct nexthop *nh,
+ *
+ * Caller must decrement ref with zebra_nhg_decrement_ref() when done.
+ */
+struct nhg_hash_entry *zebra_nhg_proto_del(uint32_t id);
```
any value in including the 'proto' here in the delete path also? it seems like it'd be a cheap way to do some level of checking that the delete is coming from the right place.

sworleys replied:
Not really, you can already determine the proto from the ID range anyway.

sworleys replied:
nvm i see what you are saying now. yes i will add a check for that.

sworleys commented on lib/hash.h:
```diff
@@ -236,7 +236,8 @@ extern void *hash_release(struct hash *hash, void *data);
  *
  * It is safe to delete items passed to the iteration function from the hash
- * table during iteration.  Please note that adding entries to the hash
+ * table during iteration. More than one item cannot be deleted during each
```
There isn't anyway to enforce it as far as I know. I added this doc here and I believe Quentin added something about it in his recent dev guide doc changes for the hash APIs as well.


There was at least one other case of hash misuse being done that quentin fixed but it was different from this.

mjstapp commented on sharpd/sharp_zebra.c:
```diff
@@ -376,14 +408,20 @@ void route_add(const struct prefix *p, vrf_id_t vrf_id,
-	for (ALL_NEXTHOPS_PTR(nhg, nh)) {
-		api_nh = &api.nexthops[i];
+	/* Only send via ID if nhgroup has been successfully installed */
+	if (nhgid && sharp_nhgroup_id_is_installed(nhgid)) {
```
mentioned this below too: this will, if I'm reading it correctly, prevent sharpd from being used to exercise backup nexthops -because zebra will unconditionally report that the group is "installed".

sworleys replied:
yea forgot to do that. Will add.

mjstapp reviewed (commented):
still have some questions, esp about the zapi and zebra side of things

sworleys commented:
Added new commit and rebased

sworleys commented:
rebased to fix conflict

mjstapp commented on zebra/zebra_nhg.h:
```diff
@@ -102,20 +102,23 @@ struct nhg_hash_entry {
- * it is unusable by our created routes but may be used by routes we get
- * from the kernel. Therefore, it is unhashable.
- */
-#define NEXTHOP_GROUP_UNHASHABLE (1 << 4)
```
how does support for user-created nhgs (using iproute e.g.) work now? for example, IGP sends a route, nhg gets created by hashing, and then someone changes the group (by id) with iproute - are we able to handle that, to notice that the group belongs to zebra and needs to match what was sent with a route?

sworleys replied:
we only track external NHGs in the table to match with routes using them. We dont actually do anything with them internally.

If we want to use a NHG thats exactly the same, it will get a new ID.

mjstapp reviewed (commented):
I appreciate the change to the zapi path - but it still looks like some sharpd functionality that we discussed will be broken by this?

sworleys commented:
> I appreciate the change to the zapi path - but it still looks like some sharpd functionality that we discussed will be broken by this?

oh yep, forgot about the backup nh changes. will do

sworleys commented on zebra/zebra_dplane.c:
```diff
@@ -2349,6 +2358,40 @@ dplane_route_update_internal(struct route_node *rn,
+		 * as that we know we have sent a nexthop group replace
+		 * or an upper level protocol has sent us the exact
+		 * same route again.
+		 */
```
this is only being short circuited if the route is using the new API. That's because with this API, the upper proto should be sending a NHG replace rather than a route replace.

tagging @donaldsharp as well, he wrote this code

sworleys commented:
> > I appreciate the change to the zapi path - but it still looks like some sharpd functionality that we discussed will be broken by this?
> 
> oh yep, forgot about the backup nh changes. will do

@mjstapp fixed this.

I added encoding/decoding functionality for backups when we are ready to use them. Right now, if zebra gets a proto NHG with a backup it early fails and then sharpd notices and reverts to the old way of sending down with route. Same as done with non-onlink nexthops.

mjstapp commented on zebra/zapi_msg.c:
```diff
@@ -1703,100 +1704,261 @@ static void zread_route_add(ZAPI_HANDLER_ARGS)
+			 * inexpensive - but it is necessary to support shared
+			 * nexthop-groups.
+			 */
+			nexthop_group_add_sorted(ng, nexthop);
```
like this, where we're testing one pointer, then using another.

mjstapp commented on sharpd/sharp_nht.c:
```diff
@@ -65,3 +69,157 @@ void sharp_nh_tracker_dump(struct vty *vty)
+	if (nhgc->backup_list_name[0])
+		bnhgc = nhgc_find(nhgc->backup_list_name);
+
+	nhg_add(snhg->id, &nhgc->nhg, (bnhgc ? &bnhgc->nhg : NULL));
```
this pattern causes the whole new nhg zapi stack to be run once for each nexthop, rather than each nhg - that seems inefficient?

sworleys replied:
Yea you want it for each nexthop so it sends a replace if someone removes/adds a nexthop. Can't do it just on group creation.

mjstapp replied:
but - that just seems ... wrong? that would mean MULTIPATH_NUM round trips for full-ecmp groups - complete round trips through zebra and the kernel. can't we identify the point where a configuration change to an nhg occurs - exiting the submode, for example - and that would support both add and modify operations?

sworleys replied:
Sure that makes sense to me. But these callbacks were already in place in lib/nexthop_group.h to subscribe to before this PR. It's not in scope to change them for this imo.

mjstapp reviewed (commented):
still saw some issues running this version, including a zebra crash - comments inline

sworleys commented:
> still saw some issues running this version, including a zebra crash - comments inline

Fixed the issues with zread_nexthops with your suggestions and added "NHG" to the ID log.


One question, do we not have any backup NHG topotests? I assumed some of this stuff would have been caught by them. I did a quick skim and did not see any. I definitely think we should have some if we really don't.

mjstapp commented:
> One question, do we not have any backup NHG topotests? I assumed some of this stuff would have been caught by them. I did a quick skim and did not see any. I definitely think we should have some if we really don't.

There are some tests of the TI-LFA work that include backups - I didn't add any tests that use sharpd to create them though. But you're right - especially if there's going to be some kind of kernel support, it'll be of more general interest ... sometime soon.

mjstapp reviewed (approved):
We've agreed to move forward with the nhg work at this point
